fms_extras/models/speculator.py

```diff
@@ -170,3 +170,94 @@ def forward(
             state = self.activation(self.ln[i](state))  # b n d
             out.append(self.head[i](state))  # b n v
         return torch.stack(out, dim=0)  # h b n v
+
+
+def select_inflate_dim(
+    inp: torch.Tensor, inds: torch.Tensor, dim: int = 0
+) -> torch.Tensor:
+    """
+    Takes an input of size ([...], n, [...]), with n in slot corresponding to value of dim,
+    and tensor of indices of size (a, ..., z). Using those indices we over/under sample the
```

Comment:
usually oversample/undersample would mean sampling with a higher/lower frequency. It looks like the code just does selection/views? if 'sampling' is correct, maybe it could be explained more clearly what's being sampled

Reply:
Generally when we use this function, we are sampling with higher or lower frequency. If `prod(inds.shape) < n`, we're undersampling, and vice versa.

Reply:
I think this is just a confusing way of describing it though... "oversampling" usually implies random sampling. we're just selecting specific items based on a pre-defined mapping. I think "over/under sampling" here is referring to whether there are more/less items in the source/destination tensors?

Reply:
Yes this is just referring to whether or not source values are necessarily being omitted or repeated in the output. I can drop the terminology to avoid the implication that the draw is random